stl/inc/limits

```diff
@@ -1009,24 +1011,96 @@ public:
-_NODISCARD constexpr int _Countr_zero(const _Ty _Val) noexcept {
-    if (_Val == 0) {
-        return numeric_limits<_Ty>::digits;
-    }
+// Implementation of countr_zero without using specialized CPU instructions.
+// Used at compile time and when said instructions are not supported.
+// see "Hacker's Delight" section 5-4
+template <class _Ty>
+_NODISCARD constexpr int _Countr_zero_helper(const _Ty _Val) noexcept {
+    constexpr int _Digits = std::numeric_limits<_Ty>::digits;
+    return _Digits - _Countl_zero_helper(static_cast<_Ty>(static_cast<_Ty>(~_Val) & static_cast<_Ty>(_Val - 1)));
+}
+
+#if defined(_M_IX86) || defined(_M_X64)
+
+extern "C" {
+extern int __isa_available;
+#ifdef __clang__
+#define _TZCNT_U32 __builtin_ia32_tzcnt_u32
+#define _TZCNT_U64 __builtin_ia32_tzcnt_u64
+#else // ^^^ __clang__ / ! __clang__ vvv
+__MACHINEX86_X64(unsigned int _tzcnt_u32(unsigned int))
+__MACHINEX64(unsigned __int64 _tzcnt_u64(unsigned __int64));
+#define _TZCNT_U32 _tzcnt_u32
+#define _TZCNT_U64 _tzcnt_u64
+#endif // __clang__
+}
 
-    if constexpr (sizeof(_Ty) <= sizeof(unsigned int)) {
-        return __builtin_ctz(_Val);
+template <class _Ty>
+int _Checked_x86_x64_countr_zero(const _Ty _Val) noexcept {
+    const bool _Have_tzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;
+    constexpr int _Digits  = numeric_limits<_Ty>::digits;
+    constexpr _Ty _Max     = (numeric_limits<_Ty>::max)();
+    if (!_Have_tzcnt && _Val == 0) {
+        return _Digits;
+    }
+
+    if constexpr (_Digits <= 32) {
+        // Intended widening to int. This operation means that a narrow 0 will widen
+        // to 0xFFFF....FFFF0... instead of 0. We need this to avoid counting all the zeros
+        // of the wider type
+        return static_cast<int>(_TZCNT_U32(static_cast<unsigned int>(~_Max | _Val)));
     } else {
-        return __builtin_ctzll(_Val);
+#ifdef _M_IX86
+        unsigned int _High = _Val >> 32;
+        unsigned int _Low  = static_cast<unsigned int>(_Val);
+        int _Result        = _Checked_x86_x64_countr_zero(_Low);
+        _Result += !_Low ? _Checked_x86_x64_countr_zero(_High) : 0;
+        return _Result;
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        return static_cast<int>(_TZCNT_U64(_Val));
+#endif // _M_IX86
     }
 }
-#endif // __cpp_lib_bitops
+#undef _TZCNT_U32
+#undef _TZCNT_U64
+#endif // defined(_M_IX86) || defined(_M_X64)
+
+template <class _Ty>
+constexpr bool _Is_standard_unsigned_integer =
+    _Is_any_of_v<remove_cv_t<_Ty>, unsigned char, unsigned short, unsigned int, unsigned long, unsigned long long>;
+
+template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> = 0>
+_NODISCARD constexpr int _Countr_zero(const _Ty _Val) noexcept {
+#ifdef __cpp_lib_is_constant_evaluated
+    if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated
+    {
+#if defined(_M_IX86) || defined(_M_X64)
+        return _Checked_x86_x64_countr_zero(_Val);
+#endif // _M_IX86

```

Comment:
Why wouldn't you use at least _BitScanForward for ARM ?

Reply:
Or is it _BitScanReverse ? Both are available

Reply:
Indeed, `intrin0.h` declares:

```
__MACHINE(unsigned char _BitScanForward(unsigned long * _Index, unsigned long _Mask))
__MACHINEX64(unsigned char _BitScanForward64(unsigned long * _Index, unsigned __int64 _Mask))
__MACHINEARM64(unsigned char _BitScanForward64(unsigned long * _Index, unsigned __int64 _Mask))
__MACHINE(unsigned char _BitScanReverse(unsigned long * _Index, unsigned long _Mask))
__MACHINEX64(unsigned char _BitScanReverse64(unsigned long * _Index, unsigned __int64 _Mask))
__MACHINEARM64(unsigned char _BitScanReverse64(unsigned long * _Index, unsigned __int64 _Mask))
```

and we're using both in `charconv`.

Reply:
I would be fine with filing a `performance` issue to use intrinsics for ARM and ARM64 here.

Reply:
yeah, one reason is that bitscanforward and bitscanreverse lead to more code. Also I couldn't find an equivalent instruction in ARMs docs (that wasn't neon) so it's not clear bitscanforward and reverse are that much better.

Reply:
Well they are better.
`clz` is count leading zeros
`rbit` reverses bit order to count trailing zeros.

https://godbolt.org/z/8QraQx

Reply:
oh I didn't know about rbit. In any case for countl_zero arm is supported and we do emit clz. I'll file this as a performance improvement bug.

Reply:
filed: https://github.com/microsoft/STL/issues/926

Reply:
Also, for ARM64, we're always allowed to assume NEON is available.
https://docs.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=vs-2019
"The ARM64 version of Windows presupposes that it's running on an ARMv8 or later architecture at all times. Both floating-point and NEON support are presumed to be present in hardware."